crates/matrix-sdk/src/notification_settings.rs

```diff
@@ -0,0 +1,1156 @@
+        // Search for an enabled Override
+        if let Some(rule) = ruleset.override_.iter().find(|x| x.enabled) {
+            // without a Notify action
+            if !rule.actions.iter().any(|x| matches!(x, Action::Notify)) {
+                // with a condition of type `EventMatch` for this room_id
+                if rule.conditions.iter().any(|x| match x {
+                    PushCondition::EventMatch { key, pattern } => {
+                        key == "room_id" && *pattern == *room_id
+                    }
+                    _ => false,
+                }) {
+                    return Some(RoomNotificationMode::Mute);
+                }
+            }
+        }
+
+        // Search for an enabled Room rule where rule_id is the room_id
+        if let Some(rule) = ruleset.room.iter().find(|x| x.enabled && x.rule_id == *room_id) {
+            // if this rule contains a Notify action
+            if rule.actions.iter().any(|x| matches!(x, Action::Notify)) {
+                return Some(RoomNotificationMode::AllMessages);
+            } else {
+                return Some(RoomNotificationMode::MentionsAndKeywordsOnly);
+            }
+        }
+
+        // There is no custom rule matching this room_id
+        None
+    }
+
+    /// Delete all user defined rules for a given room id
+    pub async fn delete_user_defined_room_rules(
+        &self,
+        room_id: &String,
+        ruleset: &mut Ruleset,
+    ) -> Result<(), NotificationSettingsError> {
+        let rules = self.get_custom_rules_for_room(room_id, ruleset);
+        self.delete_rules(rules, vec![], ruleset).await
+    }
+
+    /// Gets the default push notification mode for a given room id
+    pub fn get_default_room_notification_mode(
+        &self,
+        is_encrypted: bool,
+        members_count: u64,
+        ruleset: &Ruleset,
+    ) -> Result<RoomNotificationMode, NotificationSettingsError> {
+        // get the correct default rule id based on is_encrypted and members_count
+        let rule_id = match (is_encrypted, members_count) {
+            (true, 2) => PredefinedUnderrideRuleId::EncryptedRoomOneToOne,
+            (false, 2) => PredefinedUnderrideRuleId::RoomOneToOne,
+            (true, _) => PredefinedUnderrideRuleId::Encrypted,
+            (false, _) => PredefinedUnderrideRuleId::Message,
+        };
+
+        // If there is an Underride rule that should trigger a notification, the mode is
+        // 'AllMessages'
+        if ruleset.underride.iter().any(|r| {
+            r.enabled
+                && r.rule_id == rule_id.to_string()
+                && r.actions.iter().any(|a| a.should_notify())
+        }) {
+            Ok(RoomNotificationMode::AllMessages)
+        } else {
+            // Otherwise, the mode is 'MentionsAndKeywordsOnly'
+            Ok(RoomNotificationMode::MentionsAndKeywordsOnly)
+        }
+    }
+
+    /// Get whether the IsUserMention rule is enabled.
+    #[allow(deprecated)]
```

Comment:
```suggestion
```

Reply:
Oh, this attribute isn't just for the function definition! Thank you.